### PR TITLE
RLHF updates

### PR DESCRIPTION
**Pull request by PierpaoloSorbellini**

### CODE REVIEW

File: apps/accelerate/chatllama/chatllama/llama_model.py [hunk 1]
```diff
@@ -40,16 +97,314 @@ def __call__(self, texts: Union[List[str], str], *args, **kwargs):
...
+
+            # create left padded tokens
+            # TODO: decide how oes and bos should be handled
+            left_tokens = self.tokenizer.pad_id * torch.ones_like(tokens)
+            left_tokens[:, -1] = self.tokenizer.eos_id
+            mask = self.create_sequence_mask(tokens)
+            for i in range(tokens.shape[0]):
+                current_tokens = tokens[i, mask[i] == 1]
+                left_tokens[
+                    i, -len(current_tokens) :  # noqa E203
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/llama_model.py:109]: shouldn't this be something like 

```suggestion
                    i, -len(current_tokens) - 1 :  -1
```
so that we do not overwrite the `eos_id` token.

File: apps/accelerate/chatllama/chatllama/llama_model.py [hunk 2]
```diff
@@ -40,16 +97,314 @@ def __call__(self, texts: Union[List[str], str], *args, **kwargs):
             )
             for i, text in enumerate(texts):
                 tokens[i, : len(text)] = torch.tensor(text).cuda().long()
+
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/llama_model.py:100]: Instead of creating the left_tokens can't you just modify the line above?

Something like
```python
tokens[i, -len(text):] = torch.tensor(text).cuda().long()
```
should do the trick, right?

File: apps/accelerate/chatllama/chatllama/llama_model.py [hunk 3]
```diff
@@ -40,16 +97,314 @@ def __call__(self, texts: Union[List[str], str], *args, **kwargs):
...
+    vocab_size: int = -1
+    # make SwiGLU hidden layer size multiple of large power of 2
+    multiple_of: int = 256
+    norm_eps: float = 1e-5
+
+    max_batch_size: int = 32
+    max_seq_len: int = 1024
+
+    # added attributes
+    forze_embeddings: bool = True
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/llama_model.py:143]: `forze`?

File: apps/accelerate/chatllama/chatllama/llama_model.py [hunk 4]
```diff
@@ -59,50 +414,109 @@ class Transformer(nn.Module):
...
         return logits
 
-    def _forward(self, tokens: torch.Tensor, start_pos: int):
+    def _forward(
+        self, tokens: torch.Tensor, mask: torch.Tensor
+    ) -> torch.Tensor:
         _bsz, seqlen = tokens.shape
         h = self.tok_embeddings(tokens)
         self.freqs_cis = self.freqs_cis.to(h.device)
+        start_pos = 1  # TEMPORARY
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/llama_model.py:482]: add a short description of why this temporary fix is needed ;)

File: apps/accelerate/chatllama/chatllama/llama_model.py [hunk 5]
```diff
@@ -59,50 +414,109 @@ class Transformer(nn.Module):
...
+                if mask[i, j] != 0:
+                    attention_mask[i, j, start_pos : j + 1] = 1  # noqa E203
+        attention_mask = torch.where(
+            attention_mask == 0, float("-inf"), attention_mask
+        )
+        attention_mask = (
+            torch.where(attention_mask == 1, float(0.0), attention_mask)
+            .long()
+            .detach()
+        )
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/llama_model.py:502]:
```suggestion
        # mask has size (bsz, seqlen). It should be transformed in (bsz, seqlen, seqlen)
        # if the mask is a boolean tensor, convert it to int
        if mask.dtype == torch.bool:
            mask = mask.long()
        attention_mask = mask[:, None, None, :].expand(_bsz, 1, seqlen, seqlen)
        attention_mask = 1 - attention_mask
        attention_mask = torch.where(
            attention_mask == 1, float("-inf"), attention_mask
        ).long().detach()
```

the code above should be roughly equivalent to what you wrote. But isn't this missing the `triu` part? :)

File: apps/accelerate/chatllama/chatllama/llama_model.py [hunk 6]
```diff
@@ -177,17 +597,38 @@ def load_model(
...
+        model_args.forze_embeddings = froze_embeddings
+        model_args.use_fairscale = use_fairscale
+        tokenizer = Tokenizer(model_path=tokenizer_path)
+        model_args.vocab_size = tokenizer.n_words
+        torch.set_default_tensor_type(torch.cuda.HalfTensor)
+        model = Transformer(model_args)
+        torch.set_default_tensor_type(torch.FloatTensor)
+        model.load_state_dict(checkpoint, strict=False)
+        tokenizer = HFLikeTokenizer(tokenizer)
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/llama_model.py:624]: I think that this section should be moved somewhere else

File: apps/accelerate/chatllama/chatllama/rlhf/trainer.py [hunk 7]
```diff
@@ -9,10 +9,42 @@
...
+│                             │             │                          │
+│                             │ x Episodes  └─────────────▲────────────┘
+└───────────────┬─────────────┘                           │   x Epochs
+                │ store N Examples per Timestep           │  
+         ┌──────▼──────┐                                  │
+         │             │                                  │
+         │  Memories   ├──────────────────────────────────┘
+         │             │ (update timesteps x N Examples)
+         └─────────────┘
+"""  # noqa W291
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/rlhf/trainer.py:47]: Amazing visual explanation! 🥇

File: apps/accelerate/chatllama/chatllama/rlhf/trainer.py [hunk 8]
```diff
@@ -263,18 +296,6 @@ def __init__(
...
-            if self.deepspeed_config_path is None:
-                raise ValueError(
-                    "DeepSpeed config path is None, but deepspeed is enabled"
-                )
-            if os.path.exists(self.deepspeed_config_path) is False:
-                raise ValueError(
-                    f"DeepSpeed config path {self.deepspeed_config_path}"
-                    f"does not exist"
-                )
-
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/rlhf/trainer.py:277]: No DeepSpeed for RLHF?

File: apps/accelerate/chatllama/chatllama/llama_model.py (hunk 6 above)
Comment (PierpaoloSorbellini) [reply apps/accelerate/chatllama/chatllama/llama_model.py:624]: what do you mean?

File: apps/accelerate/chatllama/chatllama/llama_model.py (hunk 3 above)
Comment (PierpaoloSorbellini) [reply apps/accelerate/chatllama/chatllama/llama_model.py:143]: my mistake it has to be froze

File: apps/accelerate/chatllama/chatllama/llama_model.py (hunk 1 above)
Comment (PierpaoloSorbellini) [reply apps/accelerate/chatllama/chatllama/llama_model.py:109]: yep originally was exactly like this, to preserve the final EOS token.
i did change it while debugging and never revert it back. 
Anyway the fact that all the sequences must end with EOS is just assumed by their code, isn't it strange?
I mean if it has to generate the next token (the answer) why do you want a EOS in between?

File: apps/accelerate/chatllama/chatllama/rlhf/trainer.py (hunk 7 above)
Comment (PierpaoloSorbellini) [reply apps/accelerate/chatllama/chatllama/rlhf/trainer.py:47]: thanks :)

File: apps/accelerate/chatllama/chatllama/llama_model.py (hunk 5 above)
Comment (PierpaoloSorbellini) [reply apps/accelerate/chatllama/chatllama/llama_model.py:502]: This is not working as it is. 
if i have a vect [0, 0, 0, 1, 1, 1]
I should produce [
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 0, 0],
    [0, 0, 0, 1, 1, 0],
    [0, 0, 0, 1, 1, 1]
]
The output of this code will be:
I should produce [
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 1, 1, 1]
]
i can find a easy way to do this with the triu without overcomplicating this.

File: apps/accelerate/chatllama/chatllama/rlhf/trainer.py (hunk 8 above)
Comment (PierpaoloSorbellini) [reply apps/accelerate/chatllama/chatllama/rlhf/trainer.py:277]: ok the problem is the following... 
In vanilla deepspeed for what i understood the flow is the following...
(
    self.model_engine,
    self.optimizer,
    self.train_dataloader,
    _,
) = deepspeed.initialize(
    args=None,
    model=self.model,
    model_parameters=self.model.parameters(),
    training_data=self.train_dataloader,
    config=self.config.deepspeed_config_path,
)

then 
output = self.model_engine(input, mask)
loss = ...
self.model_engine.backward(loss)
self.model_engine.step()

When i train with the RLHF i have a class ActorCritic which contains two models actor and critic.
So the forward is performed as self.actorcritic.forward()
and thus i will need to do a model_engine of actorcritic.
Then the update is done separately with two different losses 
loss_actor = ... 
loss_critic = ... 
and first is updated only the actor and then only the critic so i should have a model engine for actor and critic... 
TO apply deepspeed in this case we have to rethink some classes and how are structured, maybe is best that we first try to use it vanilla with the other training and see what works and what not and then try to refactor this things to make it work also for RLHF.

File: apps/accelerate/chatllama/chatllama/rlhf/actor.py [hunk 9]
```diff
@@ -7,10 +7,11 @@
...
 from einops import rearrange
 from torch.utils.data import Dataset, DataLoader
-from config import ConfigActor
-from utils import TrainingStats
 
-from chatllama.llama_model import load_model
+from rlhf.config import ConfigActor
+from rlhf.utils import TrainingStats
+
+from llama_model import load_model
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/rlhf/actor.py:14]: please use absolute imports.
```suggestion
from chatllama.llama_model import load_model
from chatllama.rlhf.config import ConfigActor
from chatllama.rlhf.utils import TrainingStats
```

File: apps/accelerate/chatllama/chatllama/rlhf/reward.py [hunk 10]
```diff
@@ -11,9 +11,9 @@
 from transformers import GPT2Tokenizer, GPT2Model, BartModel
 from transformers import BartTokenizer, BartConfig, AutoModel, AutoTokenizer
 
-from chatllama.langchain_modules.prompt_templates import REWARD_TEMPLATE
-from config import ConfigReward
-from utils import TrainingStats
+from langchain_modules.prompt_templates import REWARD_TEMPLATE
+from rlhf.config import ConfigReward
+from rlhf.utils import TrainingStats
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/rlhf/reward.py:16]:
```suggestion
from chatllama.langchain_modules.prompt_templates import REWARD_TEMPLATE
from chatllama.rlhf.config import ConfigReward
from chatllama.rlhf.utils import TrainingStats
```

File: apps/accelerate/chatllama/chatllama/rlhf/trainer.py [hunk 11]
```diff
@@ -3,16 +3,49 @@
...
 from torch.utils.data import Dataset, DataLoader
 
-from actor import ActorModel
-from reward import RewardModel, CriticModel
-from config import ConfigReward, ConfigActor, Config
-from utils import TrainingStats, ConversationLog
+from rlhf.actor import ActorModel
+from rlhf.reward import RewardModel, CriticModel
+from rlhf.config import ConfigReward, ConfigActor, Config
+from rlhf.utils import TrainingStats, ConversationLog
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/rlhf/trainer.py:16]:
```suggestion
from chatllama.rlhf.actor import ActorModel
from chatllama.rlhf.reward import RewardModel, CriticModel
from chatllama.rlhf.config import ConfigReward, ConfigActor, Config
from chatllama.rlhf.utils import TrainingStats, ConversationLog
```

File: apps/accelerate/chatllama/chatllama/rlhf/trainer.py [hunk 12]
```diff
@@ -536,6 +623,8 @@ def train(
                 # sample num_examples examples from  example dataset
                 inputs = self.example_sampler.sample(num_examples)
 
+                print(inputs)
```
Comment (diegofiori) [inline apps/accelerate/chatllama/chatllama/rlhf/trainer.py:626]: remove print

File: apps/accelerate/chatllama/chatllama/llama_model.py (hunk 5 above)
Comment (diegofiori) [reply apps/accelerate/chatllama/chatllama/llama_model.py]: if you take the `torch.tril` matrix of what I wrote you get the correct one :)
